File: 3622-check-divisibility-by-digit-sum-and-product/3622-check-divisibility-by-digit-sum-and-product.js
/**
 * @param {number} n
 * @return {boolean}
 */
var checkDivisibility = function(n) {
    const original = n;
    let sum = 0;
    let product = 1;

    while (n > 0) {
        const digit = n % 10;

        sum += digit;
        product *= digit;

        // Force integer division in JavaScript
        n = Math.floor(n / 10);
    }

    // Edge case safeguard: division by zero is handled by returning false
    if (sum + product === 0) return false;

    return original % (sum + product) === 0;
};

File: 3622-check-divisibility-by-digit-sum-and-product/3622-check-divisibility-by-digit-sum-and-product.py
class Solution:
    def checkDivisibility(self, n: int) -> bool:
        original = n
        sum_val = 0  # Renamed from 'sum' to avoid shadowing Python's built-in sum function
        product = 1

        while n > 0:
            digit = n % 10

            sum_val += digit
            product *= digit

            n //= 10  # Enforce floor/integer division

        if sum_val + product == 0:
            return False

        return original % (sum_val + product) == 0
